external_ip_mapper.go

```diff
@@ -0,0 +1,143 @@
+package ice
+
+import (
+	"net"
+	"strings"
+)
+
+func validateIPString(ipStr string) (net.IP, bool, error) {
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return nil, false, ErrInvalidNAT1To1IPMapping
+	}
+	return ip, (ip.To4() != nil), nil
+}
+
+// ipMapping holds the mapping of local and external IP address for a particular IP family
+type ipMapping struct {
+	ipSole net.IP            // when non-nil, this is the sole external IP for one local IP assumed
+	ipMap  map[string]net.IP // local-to-external IP mapping (k: local, v: external)
+}
+
+func (m *ipMapping) setSoleIP(ip net.IP) error {
+	if m.ipSole != nil || len(m.ipMap) > 0 {
+		return ErrInvalidNAT1To1IPMapping
+	}
+
+	m.ipSole = ip
+
+	return nil
+}
+
+func (m *ipMapping) addIPMapping(locIP, extIP net.IP) error {
+	if m.ipSole != nil {
+		return ErrInvalidNAT1To1IPMapping
+	}
+
+	locIPStr := locIP.String()
+
+	// check if dup of local IP
+	if _, ok := m.ipMap[locIPStr]; ok {
+		return ErrInvalidNAT1To1IPMapping
+	}
+
+	m.ipMap[locIPStr] = extIP
+
+	return nil
+}
+
+func (m *ipMapping) findExternalIP(locIP net.IP) (net.IP, error) {
+	if m.ipSole != nil {
+		return m.ipSole, nil
+	}
+
+	extIP, ok := m.ipMap[locIP.String()]
+	if !ok {
+		return nil, ErrExternalMappedIPNotFound
+	}
+
+	return extIP, nil
+}
+
+type externalIPMapper struct {
```

Comment:
This is not my area of knowledge, but I don't think this data modeling is correct.

My main concern is why do we have an ipMapping for ipv4 and one for ipv6. The keys of both mappings should never collide, and I think it should be possible to have an external ipv4 address mapping to a local ipv6 one while it cannot happen now. If we have one mapping though we cannot have a _sole_ ipv4 and a _sole_ ipv6 address. And now I'm questioning... should we have a _sole_ ip address, or even one ipv4 and one ipv6 address at most? I think it should be possible to specify multiple ip addresses as candidates (e.g.: in a client/server scenario multiple servers are available and the client can choose the fastest one). Maybe I'm overthinking it. Feel free to ignore this comment or ping me if you want some clarification.

Reply:
Yea, I had one map for both ip families at first. Then I realized I needed to make sure (when local IP is not specified), the mapping should allow up to one IPv4 and one IPv6, and the code turned out much cleaner with the separate map... but that is all. (just an implementation detail)